Core/Src/drivers/nand_flash/nand_m79a_lld.c

```diff
@@ -439,40 +436,74 @@ NAND_ReturnType NAND_Block_Erase(PhysicalAddrs *addr) {
 
     /* Command 1: WRITE ENABLE */
     __write_enable();
-
+    NAND_ReturnType status = NAND_Wait_Until_Ready();
+    if (status != Ret_Success) {
+        return Ret_Failed;
+    }
     /* Command 2: BLOCK ERASE. See datasheet page 35 for details */
 
     // TODO: datasheet simply specifies block address. assuming that's the 11-bit
     // block number padded with dummy bits. check.
-    uint32_t block = 0x7ff & addr->block;
+    uint32_t block = (0x7ff & addr->block) << 6;
 
     uint8_t command[4] = {SPI_NAND_BLOCK_ERASE, BYTE_2(block), BYTE_1(block), BYTE_0(block)};
 
     SPI_Params tx_cmd = {.buffer = command, .length = 4};
     if (NAND_SPI_Send(&tx_cmd) != SPI_OK) {
-        return Ret_EraseFailed;
+        return Ret_Failed; // I/O failure
     }
 
     /* Command 3: wait for device to be ready again */
-    if (NAND_Wait_Until_Ready() != Ret_Success) {
-        return Ret_NANDBusy;
+    status = NAND_Wait_Until_Ready();
+    if (status != Ret_Success) {
+        NAND_Reset();
     }
-
     /* Command 4: WRITE DISABLE */
     __write_disable();
+    return status;
+}
 
-    // TODO: check erase fail bit in status register and return that
+/*
+ * False: Block is good
+ * True: Block is bad
+ */
+bool NAND_is_Bad_Block(int block) {
+    PhysicalAddrs addr = {0};
+    addr.block = block;
+    addr.column = 2048;
+    uint8_t marker = 0xFF;
+    NAND_Page_Read(&addr, sizeof(marker), &marker);
+    return marker != 0xFF;
+}
+
+NAND_ReturnType NAND_Mark_Bad_Block(int block) {
+    PhysicalAddrs addr = {0};
+    addr.block = block;
+    addr.column = 2048;
+    uint8_t marker = 0x00;
+    // It seems that sometimes we can write a couple bits to a bad block
+    // Really all the marker needs to be is not 0xFF
+    // The datasheet seems confident that this will work
+    return NAND_Page_Write(&addr, sizeof(marker), &marker);
```

Comment:
I think `NAND_Page_Write` does not exist, I had to change it in my code to `NAND_Page_Program`